lisa/sut_orchestrator/libvirt/transformers.py

```diff
@@ -0,0 +1,523 @@
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "qemu_installer_node")
+
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer,
+            node=node,
+            log=self._log,
+        )
+        qemu_version = installer.install()
+        self._log.info(f"installed qemu version: {qemu_version}")
+
+        node.reboot()
+        return {}
+
+
+class CloudHypervisorInstallerTransformer(Transformer):
+    @classmethod
+    def type_name(cls) -> str:
+        return "cloudhypervisor_installer"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return InstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: InstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "cloudhypervisor_installer_node")
+
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer,
+            node=node,
+            log=self._log,
+        )
+        ch_version = installer.install()
+        self._log.info(f"installed qemu version: {ch_version}")
+
+        node.reboot()
+        return {}
+
+
+class LibvirtPackageInstaller(BaseInstaller):
+
+    __distro_package_mapping = {
+        Ubuntu.__name__: ["libvirt-daemon-system"],
+        CBLMariner.__name__: ["dnsmasq", "ebtables", "libvirt"],
+    }
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "libvirt_package"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return PackageInstallerSchema
+
+    def validate(self) -> None:
+        assert type(self._node.os).__name__ in self.__distro_package_mapping, (
+            f"'{self.type_name()}' installer does not support {self._node.os.name} os",
+        )
+
+    def install(self) -> str:
+        node: Node = self._node
+        linux: Linux = cast(Linux, node.os)
+        packages_list = self.__distro_package_mapping[type(linux).__name__]
+        self._log.info(f"installing packages: {packages_list}")
+        linux.install_packages(list(packages_list))
+        version = self._get_version()
+        return version
+
+    def _get_version(self) -> str:
+        return _get_package_version(self._node, "libvirtd")
+
+
+class QemuPackageInstaller(BaseInstaller):
+
+    __distro_package_mapping = {
+        Ubuntu.__name__: ["qemu-kvm"],
+        CBLMariner.__name__: ["qemu-kvm"],
+    }
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "qemu_package"
```

Comment:
`package` is enough, if there is one more layer like `QemuInstaller`.

Reply:
Changed.